app/handlers/sessions_create.rb

```diff
@@ -57,7 +57,14 @@ def handle
         status = :authentication_taken
       else
         return outputs[:status] = :new_signin_required if user_signin_is_too_old?
-        run(TransferAuthentications, authentication, current_user)
+        auth_transfer = run(TransferAuthentications, authentication, current_user)
+
+        if auth_transfer.outputs[:same_provider]
```

Comment:
To me, setting `same_provider` inside `TransferAuthentications` whenever the `update_attributes` call fails is a little too indirect.  What if that `update_attributes` call failed for another validation error?  

I also got down this line of thinking because I think `TransferAuthentications` needs to `fatal_error` if it can't transfer authentications as directed.  You're right that for now there is only one caller, but we can't assume that will always be true, and using a `fatal_error` are the only way to protect a future caller from what is actually an error.

I'd rather you do a pre-check that is more explicit..

```
return outputs[:status] = :same_provider if current_user.authentications.any?{|user_auth| user_auth.provider == authentication.provider}
```

and then in `TransferAuthentications` bubble up any model errors in the normal way...

```
authentication.update_attribute(:user_id, target_user.id)
transfer_errors_from(authentication, {type: :verbatim}, true) # true for fatal
```

Leave the DB uniqueness constraint in of course.